influxql/query_executor.go

```diff
@@ -341,52 +341,63 @@ func (e *QueryExecutor) query(qid uint64) (*QueryTask, bool) {
 	return query, ok
 }
 
-// attachQuery attaches a running query to be managed by the QueryExecutor.
-// Returns the query id of the newly attached query or an error if it was
-// unable to assign a query id or attach the query to the QueryExecutor.
-// This function also returns a channel that will be closed when this
-// query finishes running.
-//
-// After a query finishes running, the system is free to reuse a query id.
-func (e *QueryExecutor) attachQuery(q *Query, database string, interrupt <-chan struct{}) (uint64, *QueryTask, error) {
-	e.mu.Lock()
-	defer e.mu.Unlock()
+// newQuery creates a new query to be managed by the QueryExecutor.
+func (e *QueryExecutor) newQuery(q *Query, database string, interrupt <-chan struct{}) (uint64, *QueryTask, error) {
+	task := &QueryTask{
+		query:     q.String(),
+		database:  database,
+		startTime: time.Now(),
+		closing:   make(chan struct{}),
+		monitorCh: make(chan error),
+	}
 
+	e.mu.Lock()
```

Comment:
I'm not certain. This was just reducing the amount of code the write lock was kept.

I can do the same thing as the tsm1 cache, but the two conditions that this checks (whether it is shutdown or max concurrent queries) are both unlikely to fail, so the normal path is for both of those checks to succeed.